### PR TITLE
rewrite expand_by_entities

### PR DESCRIPTION
**Pull request by stegben**
Drop the `include_replacement` functionality and allow exlude original entity

### CODE REVIEW

**Comment by stegben** [comment]
@SoluMilken @noobOriented Done.

File: uttut/expand_by_entities.pyx [hunk 1]
```diff
@@ -1,254 +1,100 @@
...
-
-def gen_partitioned_utterance(
-        utterance: str,
-        sorted_entities: list,  # List[Entity]
-    ) -> Tuple[List, List]:
-
+def partition_by_entities(
+        datum: Datum,
+        include_orig: bool,
+    ) -> Tuple[List[List[str]], List[Union[None, str]]]:
```
Comment (SoluMilken) [inline uttut/expand_by_entities.pyx]: Union[None, str]?

Comment (stegben) [reply uttut/expand_by_entities.pyx]: It can be a string or `None`

**Comment by stegben** [comment]
`expand_by_entities` implementation outline


- `Datum` -> `partition_by_entities` -> parts and entity names
- parts -> `sampling_methods` -> idxs of parts
- loop through idxs, then call `_aggregate_entities`

File: uttut/expand_by_entities.pyx [hunk 2]
```diff
@@ -1,254 +1,100 @@
...
         return [datum]
-    entities_with_index, parts = partition_by_entities(datum)
 
-    augmented_parts = augment_method(
-        partitioned_utterance=parts,
-        sorted_entities=entities_with_index,
-        **augment_params,
-    )
+    if sampling_method is None:
+        sampling_method = lambda k: list(range(k))  # return all possible combinations
```
Comment (SoluMilken) [inline uttut/expand_by_entities.pyx]: k?

Comment (stegben) [reply uttut/expand_by_entities.pyx]: Will change to `n_combinations`

File: uttut/tests/test_expand_by_entities.py [hunk 3]
```diff
@@ -113,28 +81,21 @@ def setUp(self):
...
                 intents=ints,
             ))
-            self.expected_data_without_replacement.append(Datum(
-                utterance=utt,
-                entities=ents_wo_re,
-                intents=ints,
-            ))
 
-    def test_correct_result_original_augmentation_method(self):
+    def test_expand_by_entiies(self):
```
Comment (SoluMilken) [inline uttut/tests/test_expand_by_entities.py]: entiies???

File: uttut/tests/test_expand_by_entities.py [hunk 4]
```diff
@@ -143,36 +104,89 @@ def test_correct_result_original_augmentation_method(self):
...
-        result = expand_by_entities(
+    fake_parts = [['a', 'b'], ['c']]
+    fake_entity_names = ['Z', None]
+
+    @patch(
+        'uttut.expand_by_entities.partition_by_entities',
+        return_value=(fake_parts, fake_entity_names),
+    )
+    @patch('uttut.expand_by_entities.get_kth_combination', return_value=['a', 'c'])
+    def test_expand_by_entities_correctly_use_sampling_method(self, patch_get_k, _):
```
Comment (SoluMilken) [inline uttut/tests/test_expand_by_entities.py]: _?

File: uttut/expand_by_entities.pyx (hunk 1 above)
Comment (SoluMilken) [reply uttut/expand_by_entities.pyx]: 原來如此

**Pull request merged**